### ISSUE TITLE
not_null dereference operator

### ISSUE THREAD

**Issue by cconverse711** [Microsoft/GSL#491]
Is there a reason why not_null has no dereference operator? It seems like the presence of `operator ->` should imply that there should be some way to dereference the `not_null<T>` to get at `*T`, just like any other pointer type. Or am I missing something?

**Comment by neilmacintosh** [comment]
None that I can recall. It's just oversight at this point. I'd be happy to look at a PR that adds `operator*`.

**Comment by bashrc-real** [comment]
#507

**Comment by neilmacintosh** [comment]
This was fixed by PR #513.

**Comment by alexey-malov** [comment]
```c++
        int ints[1] = {42};
        CustomPtr<int> p1(&ints[0]);

        using NotNull1 = not_null<decltype(p1)>;
        CHECK(*NotNull1(p1) == 42);
        *NotNull1(p1) = 43; // <=================== This will not compile
        CHECK(ints[0] == 43);
```
The PR #513 returns a copy of object instead of reference to it. code snippet mentioned above illustrates the issue.

I added a PR #516 that fixes this by declaring return value of not_null::operator* as decltype(auto)
